src/ccstruct/image.h

```diff
@@ -33,6 +33,8 @@ class TESS_API Image {
   // service
   operator Pix *() const { return pix_; }
   Pix *operator->() const { return pix_; }
+  Pix **a() { return &pix_; }
```

Comment:
operator Pix**() ?

Reply:
It not really operator. It's a function call, but it acts like `&` operator.

Reply:
I mean, `a()` is very very bad.
`operator Pix**()` must be used instead.

Reply:
Yes, I know it's not nice. I didn't know how to implement it using operator overloading. I don't write C++ code very often, so excuse me.

Reply:
Try to replace it with `operator Pix**() const { return &pix_; }` and check if it works.

Reply:
```
/../src/ccstruct/image.h:37:35: error: invalid conversion from ‘Pix* const*’ to ‘Pix**’ [-fpermissive]
   37 |   operator Pix**() const { return &pix_; }
      |                                   ^~~~~
      |                                   |
      |                                   Pix* const*

```

Reply:
Try `operator const Pix**`.

Reply:
Ok, this should work `operator Pix **() const { return (Pix **)&pix_; }`.

Reply:
```
./../src/ccstruct/image.h: In member function ‘tesseract::Image::operator const Pix**() const’:
./../src/ccstruct/image.h:39:41: error: invalid conversion from ‘Pix* const*’ to ‘const Pix**’ [-fpermissive]
   39 |   operator const Pix**() const { return &pix_; }
      |                                         ^~~~~
      |                                         |
      |                                         Pix* const*

```

Reply:
This is better `operator Pix **() { return &pix_; }` since we can modify `pix_` after return.

Reply:
with
`operator Pix* const*() const { return &pix_; }`

I don't get an error in `image.h`, but I get errors from other places.

Reply:
```
/../src/ccmain/thresholder.cpp: In member function ‘void tesseract::ImageThresholder::SetImage(tesseract::Image)’:
./../src/ccmain/thresholder.cpp:157:12: error: ambiguous overload for ‘operator!=’ (operand types are ‘tesseract::Image’ and ‘std::nullptr_t’)
  157 |   if (pix_ != nullptr) {
      |       ~~~~~^~~~~~~~~~
./../src/ccmain/thresholder.cpp:157:12: note: candidate: ‘operator!=(Pix**, Pix**)’ <built-in>
./../src/ccmain/thresholder.cpp:157:12: note: candidate: ‘operator!=(Pix*, Pix*)’ <built-in>
./../src/ccmain/thresholder.cpp: In member function ‘virtual std::tuple<bool, tesseract::Image, tesseract::Image, tesseract::Image> tesseract::ImageThresholder::Threshold(tesseract::ThreshMethod)’:
./../src/ccmain/thresholder.cpp:217:34: error: cannot convert ‘tesseract::Image*’ to ‘PIX**’ {aka ‘Pix**’}
  217 |                                  &pix_thresholds, &pix_binary);
      |                                  ^~~~~~~~~~~~~~~
      |                                  |
      |                                  tesseract::Image*
In file included from ./../src/ccmain/thresholder.cpp:19:
/usr/include/leptonica/allheaders.h:152:146: note:   initializing argument 7 of ‘l_ok pixOtsuAdaptiveThreshold(PIX*, l_int32, l_int32, l_int32, l_int32, l_float32, PIX**, PIX**)’
  152 | LEPT_DLL extern l_ok pixOtsuAdaptiveThreshold ( PIX *pixs, l_int32 sx, l_int32 sy, l_int32 smoothx, l_int32 smoothy, l_float32 scorefract, PIX **ppixth, PIX **ppixd );
      |                                                                                                                                            ~~~~~~^~~~~~

```

Reply:
I've pushed proper method. Please rebase.

Reply:
I get these errors:

```
./src/ccmain/thresholder.cpp: In member function ‘virtual std::tuple<bool, tesseract::Image, tesseract::Image, tesseract::Image> tesseract::ImageThresholder::Threshold(tesseract::ThreshMethod)’:
./../src/ccmain/thresholder.cpp:217:34: error: cannot convert ‘tesseract::Image*’ to ‘PIX**’ {aka ‘Pix**’}
  217 |                                  &pix_thresholds, &pix_binary);
      |                                  ^~~~~~~~~~~~~~~
      |                                  |
      |                                  tesseract::Image*
In file included from ./../src/ccmain/thresholder.cpp:19:
/usr/include/leptonica/allheaders.h:152:146: note:   initializing argument 7 of ‘l_ok pixOtsuAdaptiveThreshold(PIX*, l_int32, l_int32, l_int32, l_int32, l_float32, PIX**, PIX**)’
  152 | LEPT_DLL extern l_ok pixOtsuAdaptiveThreshold ( PIX *pixs, l_int32 sx, l_int32 sy, l_int32 smoothx, l_int32 smoothy, l_float32 scorefract, PIX **ppixth, PIX **ppixd );
      |                                                                                                                                            ~~~~~~^~~~~~
./../src/ccmain/thresholder.cpp:219:63: error: cannot convert ‘tesseract::Image*’ to ‘PIX**’ {aka ‘Pix**’}
  219 |     r = pixSauvolaBinarizeTiled(pix_grey, 25, 0.40, 300, 300, &pix_thresholds,
      |                                                               ^~~~~~~~~~~~~~~
      |                                                               |
      |                                                               tesseract::Image*
In file included from ./../src/ccmain/thresholder.cpp:19:
/usr/include/leptonica/allheaders.h:155:123: note:   initializing argument 6 of ‘l_ok pixSauvolaBinarizeTiled(PIX*, l_int32, l_float32, l_int32, l_int32, PIX**, PIX**)’
  155 | LEPT_DLL extern l_ok pixSauvolaBinarizeTiled ( PIX *pixs, l_int32 whsize, l_float32 factor, l_int32 nx, l_int32 ny, PIX **ppixth, PIX **ppixd );
      |                                                                                                                     ~~~~~~^~~~~~
make[2]: *** [Makefile:5704: src/ccmain/libtesseract_la-thresholder.lo] Error 1
```

Reply:
You need `&pix_thresholds` -> `*pix_thresholds`.